File: part1.cpp
#include <bits/stdc++.h>
using namespace std;

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

// selection sort function
void selectionSort(int arr[], int n) {
    int i, j, min_idx;
    for(i = 0; i < n-1; i++){
        min_idx = i; 
        for (j = i+1; j < n; j++) 
        if (arr[j] < arr[min_idx]) 
            min_idx = j; 
        swap(&arr[min_idx], &arr[i]); 
    }

    
    
    // print the sorted array
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    cout << endl;
}

// bubble sort function
void bubbleSort(int arr[], int n) {
     int i, j; 
    for (i = 0; i < n-1; i++){     
        for (j = 0; j < n-i-1; j++) {
            if (arr[j] > arr[j+1]) {
                swap(&arr[j], &arr[j+1]); 
            }
        }
    }
    // print the sorted array
    for(i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    cout << endl;
}

// insertion sort function 
void insertionSort(int arr[], int n) {
    int i,key,j; 
    for(i = 1; i < n; i++){
        key = arr[i];
        j = i - 1;

        while(j>=0 && arr[j] > key) {
            arr[j+1]=arr[j];
            j=j-1;
        }
        arr[j+1]=key;
    }

    // print the sorted array
    for(i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }

    cout << endl;
    
}


// quick sort function
int partition(int a[], int start, int end) {
    int pivot = a[start];
    int count = 0;
    for (int i = start + 1; i <= end; i++) {
        if (a[i] < pivot) {
            count++;
        }
    }
    int pivotIndex = start + count;
    swap(&a[start], &a[pivotIndex]);
    int i = start;
    int j = end;
    while (i<pivotIndex && j>pivotIndex) {
        while (a[i] < pivot) {
            i++;
        }
        while (a[j] > pivot) {
            j--;
        }
        if (i<=j) {
            swap(&a[i], &a[j]);
            i++;
            j--;
        }

    }
    return pivotIndex;

}

void quickSort(int a[], int start, int end) {
    if (start < end) {
        int pivotIndex = partition(a, start, end);
        quickSort(a, start, pivotIndex - 1);
        quickSort(a, pivotIndex + 1, end);
    }

    // print the sorted array
    for(int i = 0; i < end; i++) {
        cout << a[i] << " ";
    }

    cout << endl;
}

// merge sort function
void merge(int array[], int const left, int const middle, int const right) {
    auto const subarrayone = middle - left + 1;
    auto const subarraytwo = right - middle;
    auto *leftsubarray = new int[subarrayone];
    auto *rightsubarray = new int[subarraytwo];
    for (auto i = 0; i < subarrayone; i++) {
        leftsubarray[i] = array[left + i];
    }
    for (auto i = 0; i < subarraytwo; i++) {
        rightsubarray[i] = array[middle + 1 + i];
    }
    auto i = 0;
    auto j = 0;
    auto k = left;
    while (i < subarrayone && j < subarraytwo) {
        if (leftsubarray[i] <= rightsubarray[j]) {
            array[k] = leftsubarray[i];
            i++;
        } else {
            array[k] = rightsubarray[j];
            j++;
        }
        k++;
    }
    while (i < subarrayone) {
        array[k] = leftsubarray[i];
        i++;
        k++;
    }
    while (j < subarraytwo) {
        array[k] = rightsubarray[j];
        j++;
        k++;
    }
}

void mergeSort(int array[], int const left, int const right) {
    if (left < right) {
        auto const middle = left + (right - left) / 2;
        mergeSort(array, left, middle);
        mergeSort(array, middle + 1, right);
        merge(array, left, middle, right);
    }

    // print the sorted array
    for(int i = 0; i < right; i++) {
        cout << array[i] << " ";
    }

    cout << endl;
}



int main(int argc, char const *argv[])
{
    int n;
    cout<<"Enter Size: ";
    cin >> n;
    cout<<endl;
    int arr[n];
    cout<<"Enter Array LOL XD: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    cout<<endl;
    cout<<"Selection Sort: "<<endl;
    selectionSort(arr, n);
    cout<<"Bubble Sort: "<<endl;
    bubbleSort(arr, n);
    cout<<"Insertion Sort: "<<endl;
    insertionSort(arr, n);
    cout<<"Quick Sort: "<<endl;
    quickSort(arr, 0, n-1);
    cout<<"Merge Sort: "<<endl;
    mergeSort(arr, 0, n-1);

    return 0;
}

File: part1.java
/**
 * part1
 */
import java.util.*;

public class part1 {

  public static int newArr[] = new int[10];

  public static void main(String[] args) {
    SortingAlgos a = new SortingAlgos();
    Random rand = new Random();

    for (int i = 0; i < 10; i++) {
      newArr[i] = rand.nextInt(100);
    }

    System.out.println("Unsorted Array: ");
    for (int i = 0; i < 10; i++) {
      System.out.print(newArr[i] + " ");
    }

    a.selectionSorted(newArr);
    System.out.println("Selection Sort: ");
    a.printArray(newArr);
    a.bubbleSort(newArr);
    System.out.println("Bubble Sort: ");
    a.printArray(newArr);
    a.insertionSort(newArr);
    System.out.println("Insertion Sort: ");
    a.printArray(newArr);
    QuickSort quickSort = new QuickSort();
    quickSort.quickSort(newArr, 0, newArr.length - 1);
    System.out.println("Quick Sort: ");
    a.printArray(newArr);
    quickSort.quickSortComparator();
    MergeSort mergeSort = new MergeSort();
    mergeSort.mergeSort(newArr, 0, newArr.length - 1);
    System.out.println("Merge Sort: ");
    a.printArray(newArr);

    System.out.println("\n\n");
    System.out.println("Sorted Array: ");
    for (int i = 0; i < 10; i++) {
      System.out.print(newArr[i] + " ");
    }
    System.out.println("\n\n");
    PartB partB = new PartB();
    partB.doover(newArr);
  }

  public int[] getArray() {
    return newArr;
  }
}

class PartB {

  public void doover(int newArr[]) {
    SortingAlgos a = new SortingAlgos();
    a.selectionSorted(newArr);
    System.out.println("Selection Sort: ");
    a.printArray(newArr);
    a.bubbleSort(newArr);
    System.out.println("Bubble Sort: ");
    a.printArray(newArr);
    a.insertionSort(newArr);
    System.out.println("Insertion Sort: ");
    a.printArray(newArr);
    QuickSort quickSort = new QuickSort();
    quickSort.quickSort(newArr, 0, newArr.length - 1);
    System.out.println("Quick Sort: ");
    a.printArray(newArr);
    quickSort.quickSortComparator();
    MergeSort mergeSort = new MergeSort();
    mergeSort.mergeSort(newArr, 0, newArr.length - 1);
    System.out.println("Merge Sort: ");
    a.printArray(newArr);
  }
}

class SortingAlgos {

  public void selectionSorted(int arr[]) {
    int n = arr.length;
    int comp = 0;
    int swap = 0;
    for (int i = 0; i < n - 1; i++) {
      int min_index = i;
      for (int j = 0; j < n; j++) {
        if (arr[j] < arr[min_index]) {
          min_index = j;
        }
      }

      int temp = arr[min_index];
      arr[min_index] = arr[i];
      arr[i] = temp;
    }

    for (int i = 0; i < arr.length - 1; i++) {
      int min = i;
      for (int j = i + 1; j < arr.length; j++) {
        if (arr[j] < arr[min]) {
          comp++;
        }

        if (min != i) {
          swap++;
        }
      }
    }
    System.out.println("Selection Sort Comparisons: " + comp);
    System.out.println("Selection Sort Swaps: " + swap);
  }

  public void bubbleSort(int arr[]) {
    int n = arr.length;
    int comp = 0;
    for (int i = 0; i < n - 1; i++) for (int j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        int temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
        comp++;
      }
    }
    System.out.println("Bubble Sort Comparisons: " + comp);
  }

  public void insertionSort(int arr[]) {
    int n = arr.length;
    int count = 0;
    for (int i = 0; i < arr.length; ++i) {
      int key = arr[i];
      int j = i - 1;

      while (j >= 0 && arr[j] > key) {
        arr[j + i] = arr[j];
        j = j - 1;
      }
      arr[j + 1] = key;
      count++;
    }
    System.out.println("Insertion Sort Comparisons: " + count);
  }

  public void printArray(int arr[]) {
    int n = arr.length;
    for (int i = 0; i < n; ++i) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }
}

class QuickSort {

  private static int compCounter, swapCounter;

  public void quickSort(int arr[], int begin, int end) {
    if (begin < end) {
      int partitionIndex = partition(arr, begin, end);

      quickSort(arr, begin, partitionIndex - 1);
      quickSort(arr, partitionIndex + 1, end);
    }
  }

  private int partition(int arr[], int begin, int end) {
    int pivot = arr[end];
    int i = (begin - 1);

    for (int j = begin; j < end; j++) {
      compCounter++;
      if (arr[j] <= pivot) {
        i++;

        int swapTemp = arr[i];
        arr[i] = arr[j];
        arr[j] = swapTemp;
        swapCounter++;
      }
    }
    swapCounter++;
    int swapTemp = arr[i + 1];
    arr[i + 1] = arr[end];
    arr[end] = swapTemp;

    return i + 1;
  }

  public void quickSortComparator() {
    System.out.println("quickSortComparator: " + compCounter);
  }
}

class MergeSort {

  void merge(int array[], int p, int q, int r) {
    int n1 = q - p + 1;
    int n2 = r - q;

    int L[] = new int[n1];
    int M[] = new int[n2];

    for (int i = 0; i < n1; i++) L[i] = array[p + i];
    for (int j = 0; j < n2; j++) M[j] = array[q + 1 + j];

    int i, j, k;
    i = 0;
    j = 0;
    k = p;

    while (i < n1 && j < n2) {
      if (L[i] <= M[j]) {
        array[k] = L[i];
        i++;
      } else {
        array[k] = M[j];
        j++;
      }
      k++;
    }

    while (i < n1) {
      array[k] = L[i];
      i++;
      k++;
    }

    while (j < n2) {
      array[k] = M[j];
      j++;
      k++;
    }
  }

  void mergeSort(int array[], int left, int right) {
    if (left < right) {
      int mid = (left + right) / 2;
      mergeSort(array, left, mid);
      mergeSort(array, mid + 1, right);
      merge(array, left, mid, right);
    }
  }
}
